### ISSUE TITLE
Support multiple outputs in gcc depfile

### ISSUE THREAD

**Issue by jpakkane** [ninja-build/ninja#1184]
```
ninja: error: build.ninja:233: multiple outputs aren't (yet?) supported by depslog; bring this up on the mailing list if it affects you
```

This has been around a long time [and is known to be difficult](https://groups.google.com/forum/#!searchin/ninja-build/depslog%7Csort:relevance/ninja-build/4RK3vACeHG0/T3iEPSoLEAAJ). However this is becoming more and more acute. There are many compilers that produce multiple files for each input, for example Fortran produces `.o` and `.mod`. Some compilers want to compile many files at once, e.g. `somecomp -c foo1.src foo2.src foo3.src` which produces three object files in one go.

In [Meson](https://github.com/mesonbuild/meson) we are getting reports from users who have custom compile steps such as converting one IDL file to multiple source files where this fails because depslog can have deps only for one output file.

Please add this functionality. Thanks.

**Comment by nico** [comment]
It's a valid feature request, and we should figure out how to do this. In the meantime, maybe meson could add `deps =` lines to edges with more than one output.

**Comment by mikkelfj** [comment]
I am experimenting with meson fo use with in IDL compiler (flatcc for flatbuffers). The tool generates a single depfile, currently only with one target, but in reality it has multiple targets: <schema>_reader.h, <schema>_builder.h etc., but furthermore there are included subschema <subschema1>_reader.h etc. It suffice to depend on the top-level <schema>_reader.h for correct builds usually, but the ninja clean reports errors with subschema. The is a related issue about ninja not supporting dynamically generated output such as <subschema1>_reader.h, even if it supports dynamically generated dependencies, so I'm not sure if it is that simple to get clean working.

**Comment by evmar** [comment]
"ninja clean" can only delete files it knows about; if at all possible, I recommend putting your build output in a different directory so cleaning is as easy as `rm -rf build`.

**Comment by mikkelfj** [comment]
Wrt. separate build dir, that already happens, so it is not a showstopper.

**Comment by jpakkane** [comment]
> In the meantime, maybe meson could add deps = lines to edges with more than one output.

If we do that Ninja will immediately halt saying that dep file has more than one output. We simply can't enable it until this is fixed. Nobody would be able to compile code with that.

For more details see [this bug in Gnome's bugzilla](https://bugzilla.gnome.org/show_bug.cgi?id=774368).

**Comment by drothlis** [comment]
There is a subset of this behaviour that might be simpler to fix: A depfile with additional outputs that don't have any dependencies. Like this:

```
res.xml: res1.dat res2.dat
res1.dat:
res2.dat:
```

Compilers output these additional targets for each of the dependencies so that `make` won't complain if you *remove* one of these implicit dependencies. However Ninja already handles missing dependencies so these additional targets aren't strictly necessary. The solution for this case would be to ignore targets (in the depfile) with no dependencies.

This solution would seem to handle the Gnome bug that @jpakkane mentioned in the previous comment.

C pre-processors are configurable enough that you can turn off those phony targets (GCC only produces them if you add `-MP`). However other types of compilers might not have this option (as in @jpakkane's case).

**Comment by greneholt** [comment]
For anyone that is being blocked by this issue, you may find these pull requests and patches helpful:
https://github.com/ninja-build/ninja/pull/1236
https://github.com/moroten/ninja/tree/multiple-outputs-with-deps

**Comment by mathstuf** [comment]
#1534 should have fixed this (and is in 1.10.0).
